enhancements/etcd/etcd-encryption-for-separate-oauth-apis.md

```diff
@@ -0,0 +1,149 @@
+
+
+## Design Details
+
+### Test Plan
+
+To validate if `CAO` is capable of managing its own `encryption-config` we are going to create the following E2E tests:
+ - scenario 1: turn on encryption
+ - scenario 2: turn encryption on and off
+ - scenario 3: turn on encryption and force key rotation
+ - scenario 4: measure migration performance
+  
+Note: the above tests will be created based on the common test library that drives the same set of tests for `OAS-O` and `KAS-O`.
+
+To validate upgrade / downgrade path for `4.n` and `4.(n-1)` we are going to create the following E2E tests:
+ - scenario 1: take `4.(n-1)` cluster, encrypt it, upgrade to `4.n`, downgrade and upgrade again
+ - scenario 2: take `4.(n-1)` cluster, encrypt it, upgrade to `4.n`, force key rotation, downgrade to `4.(n-1)`, force key rotation and upgrade again
+ 
+Note: At the moment we don't have tests like that so creating them will be significantly harder.
+
+To validate upgrade / downgrade path for future release `4.(n+1)` we are going to create the following E2E tests:
+ - scenario 1: take `4.n` cluster, encrypt it but let `OAS-O` manage `encryption-config`, upgrade to the next version in which `CAO` will take over the config then downgrade
+ - scenario 2: take `4.n` cluster, encrypt it but let `OAS-O` manage `encryption-config`, upgrade, force key rotation, downgrade, force key rotation and upgrade again
+
+##### Removing a deprecated feature
+
+See upgrade/downgrade.
+
+### Upgrade / Downgrade Strategy for `4.n` and `4.(n-1)`
+
+In an upgrade case, `OAS-O` in version `4.n` will be responsible for synchronizing and maintaining encryption state for both `openshift-apiserver` and `oauth-apiserver`.
+Since both will share exactly the same `encryption-config`, the new component will be able to read (decrypt) already encrypted data.
+
+During an upgrade, the new `UnionRevisionLabelPodDeployer` will create back pressure in the system and will make the encryption controllers wait for the new component to synchronize. 
+In that state, the status of `OAS-O` won't change and we won't roll out any new encryption keys.
+
+In a downgrade scenario, `OAS-O` in version `4.(n-1)` is responsible for synchronizing and maintaining encryption state only for `openshift-apiserver`.
+Since in version `4.n` exactly the same `encryption-config` was used the `openshift-apiserver` will be able to read (decrypt) data. 
+
+
+### Upgrade / Downgrade Strategy for `4.(n+1)` and `4.n`
+
+In an upgrade case, `CAO` in version `4.(n+1)` takes over `encryption-config-openshift-ouath` by removing the annotation and copying the encryption keys (`encryption-key-openshift-apiserver-{0,1, ...}`).
+From that point it will manage its own configuration.
+
+In an downgrade scenario nothing changes because `OAS-O` in version `4.n` will not manage `encryption-config-openshift-ouath` as it doesn't have the annotation. 
+Additionally `CAO` in version `4.n` was prepared to take care of its own configuration.
+
```

Comment:
would be more clear: in 4.n we have two modes of operations

1) a cluster that has never been at `4.(n+1)`
2) a cluster that has.